### PR TITLE
Adding plotting metrics to WandB

### PR DESCRIPTION
**Pull request by IFenton**
Working on adding some basic metrics to the WandB plots. Initially RMSE, MAE and SIEError, plotted by by forecast day and as an overall mean

### CODE REVIEW

File: icenet_mp/callbacks/metric_summary_callback.py [hunk 1]
```diff
@@ -53,9 +57,37 @@ def on_test_epoch_end(
...
+            if values.numel() > 1:
+                table = wandb.Table(
+                    data=list(enumerate(values.tolist(), start=1)),
+                    columns=["day", name],
+                )
+                plot_name = name + " per day"
+                wandb.log(
+                    {plot_name: wandb.plot.line(table, "day", name, title=plot_name)}
+                )
+
```
[the remark this answers is not in the document]
Comment (IFenton) [reply icenet_mp/callbacks/metric_summary_callback.py]: Done in [23fc727](https://github.com/alan-turing-institute/icenet-mp/pull/216/commits/23fc727bffd10caf2441b389d5c4060658fca071)

File: icenet_mp/models/metrics/sie_error_abs.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+"""SIEError metric."""
+
+import torch
+from torchmetrics import Metric
+
+SEA_ICE_THRESHOLD = 0.15  # Threshold for defining sea ice extent
+
+
+class SIEErrorDaily(Metric):
```
[the remark this answers is not in the document]
Comment (IFenton) [reply icenet_mp/models/metrics/sie_error_abs.py]: I was using the `Daily` tag to indicate that these metrics produce one value for each forecast day. So I've got the same naming in RMSE and MAE. Would `SeaIceExtentErrorDaily` be a reasonable name?

File: icenet_mp/models/metrics/base_metrics.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            Ignored (present for API compatibility).
+
+        """
+        rmse = torch.sqrt(torch.mean((preds - target) ** 2, dim=(0, 2, 3, 4)))
+        if self.rmse_daily.numel() == 0:
+            self.rmse_daily = rmse.unsqueeze(1)  # Shape: (T,)
+        else:
+            self.rmse_daily = torch.cat(
+                (self.rmse_daily, rmse.unsqueeze(1)), dim=1
+            )  # Shape: (T, N)
```
[the remark this answers is not in the document]
Comment (IFenton) [reply icenet_mp/models/metrics/base_metrics.py]: Done in [5ebb7d0](https://github.com/alan-turing-institute/icenet-mp/pull/216/commits/5ebb7d0e9c36ca133a108a77b5d457080842cb18)

File: icenet_mp/models/base_model.py [hunk 4]
```diff
@@ -139,6 +146,16 @@ def test_step(
...
+        # update test metrics with the current batch; computation will be done at epoch end
+        self.test_metrics.update(prediction, target)
+
+        # log test metrics aggregated over the entire test epoch to avoid per-step compute overhead
+        for name, value in self.test_metrics.items():
+            if isinstance(value, torch.Tensor):
+                self.log(
+                    name, value.mean(), on_step=False, on_epoch=True, prog_bar=False
+                )
+
```
[the remark this answers is not in the document]
Comment (IFenton) [reply icenet_mp/models/base_model.py]: Turns out that bit of code was never actually called. All that should happen in the `base_model` is that the metrics are updated with each batch (by calling `.update` on the metric list). The logging happens in the `metric_summary_callback` in `on_test_end`. That is when compute() it called, which returns both the individual values for each forecast day and the mean value across all days

**Review by jemrobinson — CHANGES_REQUESTED** [review]
There are a few open questions/comments here. I also think we can remove the `wandb_metric_callback` and further simplify the `metric_summary_callback` so that its only purpose is to create these per-forecast-day metric plots. I think everything else will be handled by the metrics themselves.

File: icenet_mp/callbacks/metric_summary_callback.py [hunk 5]
```diff
@@ -53,9 +58,43 @@ def on_test_epoch_end(
         # Post-process accumulated metrics into a single value
         metrics_: dict[str, float] = {}
         for name, values in self.metrics.items():
-            if name.startswith("average_"):
+            if name.startswith("average_") and values:
                 metrics_[name] = statistics.mean(values)
```
[the remark this answers is not in the document]
Comment (IFenton) [reply icenet_mp/callbacks/metric_summary_callback.py]: Done in [df05641](https://github.com/alan-turing-institute/icenet-mp/pull/216/commits/df05641013aecc46d12a87bbde8952cc1aeb1f5c)

File: icenet_mp/metrics/sie_error_abs.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
+"""SIEError metric."""
+
+import torch
+from torchmetrics import Metric
+
+SEA_ICE_THRESHOLD = 0.15  # Threshold for defining sea ice extent
+
+
+class SIEErrorDaily(Metric):
```
[the remark this answers is not in the document]
Comment (IFenton) [reply icenet_mp/metrics/sie_error_abs.py]: Done in [eb665b5](https://github.com/alan-turing-institute/icenet-mp/pull/216/commits/eb665b5c9adda0e1ff4e582b87b62907dd26be23)

File: icenet_mp/callbacks/metric_summary_callback.py [hunk 7]
```diff
@@ -53,9 +58,43 @@ def on_test_epoch_end(
...
+
+    def on_test_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        """Called at the end of testing."""
+        test_metrics: MetricCollection = pl_module.test_metrics  # type: ignore[assignment]
+        if not hasattr(test_metrics, "items"):
+            logger.warning(
+                "test_metrics does not have an items() method, skipping metric summary."
+            )
+            return
+
```
[the remark this answers is not in the document]
Comment (IFenton) [reply icenet_mp/callbacks/metric_summary_callback.py]: Done in [637fd6b](https://github.com/alan-turing-institute/icenet-mp/pull/216/commits/637fd6b1f2c41a7e2df53f65ac15363f87c25811)
